<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Modularized and interactive analyses over a FacileDataStore • FacileAnalysis</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><link href="extra.css" rel="stylesheet">
<meta property="og:title" content="Modularized and interactive analyses over a FacileDataStore">
<meta property="og:description" content='Provides modularized analysis methods commonly performed over 
    genomics data. Analysis results can be interactively explored
    by analysts using the "viz" or "shine" functions, which produce JavaScript-
    and shiny-based visualizations. Results can also be embedded into
    Rmarkdown reports using their respective "report" functions.'>
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">FacileAnalysis</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.98.15</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/FacileAnalysis-RNAseq.html">RNA-seq Analysis: The Facile Way</a>
    </li>
    <li>
      <a href="articles/FacileAnalysis-examples.html">Reproducing genomic anlayses the facile way</a>
    </li>
    <li>
      <a href="articles/FacileAnalysis-fdge.html">Facile Differential Expression Analysis</a>
    </li>
    <li>
      <a href="articles/FacileAnalysis-fpca.html">Facile Principal Components Analysis</a>
    </li>
    <li>
      <a href="articles/FacileAnalysis-fsea.html">Facile (Gene) Set Expression Analysis</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/facilebio/FacileAnalysis/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
<div class="section level1">
<div class="page-header"><h1 id="facileanalysis">FacileAnalysis<a class="anchor" aria-label="anchor" href="#facileanalysis"></a>
</h1></div>
<!-- badges: start -->

<p>The FacileAnalysis package defines a set of analysis tasks over genomic data in a modular fashion, which requires these data to be stored in a container that implements the FacileData API (aka a <code>FacileDataStore</code>).</p>
<p>The over-arching goal of developing analyses within this framework is to enable quick and effortless interrogation of genomic data by enabling a <strong>hybrid</strong> interactive and code-driven approach to data analysis. Analyses can be either completely code-driven, GUI driven, or some mix of the two.</p>
<p>To achieve this goal, analysis modules break down a general analysis task into smaller constituent steps, the results of which can:</p>
<ol style="list-style-type: decimal">
<li>be piped together (<code>%&gt;%</code>) to perform a complete analysis;</li>
<li>be explored at <strong>different levels of interactivity</strong>, via the <code><a href="reference/FacileAnalysisResultViz.html">shine()</a></code>, <code><a href="reference/FacileAnalysisResultViz.html">viz()</a></code>, and <code><a href="reference/FacileAnalysisResultViz.html">report()</a></code> methods; and</li>
<li>act as starting (or reference) points for <strong>the next</strong> analysis task, either directly or by via the <code><a href="reference/ranks_and_signatures.html">ranks()</a></code>, <code><a href="reference/signature.html">signature()</a></code>, or <code><a href="reference/compare.html">compare()</a></code> methods.</li>
</ol>
<p>Please refer to the <a href="https://facilebio.github.io/FacileAnalysis/articles/FacileAnalysis-RNAseq.html" class="external-link">RNA-seq Analysis: The Facile Way</a> to get an idea of how to use these tools when analyzing RNA-seq data.</p>
<!--
You can find a more in-depth overview and philosophy of this package in the
[FacileAnalysis vignette][pkgvign] (this is incomplete).
-->
<p>The analyses implemented within this package are listed below, with links to vignettes that describe their functionality in more detail:</p>
<ul>
<li>
<code>fpca</code>: <a href="FacileAnalysis-fpca.html">Interactive Principal Components Analyses</a>
</li>
<li>
<code>fdge</code>: <a href="FacileAnalysis-fdge.html">Interactive Differential Gene Expression Analysis</a>
</li>
<li>
<code>fsea</code>: <a href="FacileAnalysis-fsea.html">Interactive (Gene) Set Enrichment Analysis</a>
</li>
</ul>
<p><strong>A note on the experimental lifecycle</strong>: This package is tagged as “experimental” due to its limited use by a broader audience, and not as a sign of the commitment to its development or how long it has been in (internal) use. As we find the edge cases and pain points in the APIs through broader adoption, we expect to soon move to “maturing” (and eventually “stable”) <a href="https://www.tidyverse.org/lifecycle/" class="external-link">lifecycle</a>.</p>
<div class="section level2">
<h2 id="example-analysis-differential-expression">Example Analysis: Differential Expression<a class="anchor" aria-label="anchor" href="#example-analysis-differential-expression"></a>
</h2>
<p>We’ll include an exemplar differential expression analysis here in order to touch on some of the guiding principles of an analysis modules mentioned above.</p>
<p>We have first defined a complete differential expression analysis (<code>fdge</code>), by breaking it down into the following steps:</p>
<ol style="list-style-type: decimal">
<li>identifying the subset of samples from a <code>FacileDataStore</code> to perform the analysis over;</li>
<li>defining the model (design) over the samples, ie. the covariate to test and which covariates to use to control batch effects</li>
<li>defining the parameters to run the statistical test, ie.
<ol style="list-style-type: decimal">
<li>the assay from the samples to run the analysis on</li>
<li>the differential expression pipeline to use (<code>"voom"</code>, <code>"edgeR-qlf"</code>, <code>"limma-trend"</code>, <code>"limma"</code>)</li>
<li>advanced options, like the threshold to test against (limma/treat), or whether to incorporate sample level weights</li>
</ol>
</li>
</ol>
<p>The example below identifies genes differentially expression between tumor and normal samples (and controlling for sex) in the “BLCA” indication of the example TCGA dataset included in the <a href="https://facilebio.github.io/FacileData" class="external-link">FacileData</a> package.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/facilebio/FacileData" class="external-link">FacileData</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/facilebio/FacileAnalysis" class="external-link">FacileAnalysis</a></span><span class="op">)</span>
<span class="va">efds</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/FacileData/man/exampleFacileDataSet.html" class="external-link">exampleFacileDataSet</a></span><span class="op">(</span><span class="op">)</span>

<span class="co"># Step 1: define the samples implicated in our test</span>
<span class="va">samples</span> <span class="op">&lt;-</span> <span class="fu">filter_samples</span><span class="op">(</span><span class="va">efds</span>, <span class="va">indication</span> <span class="op">==</span> <span class="st">"BLCA"</span><span class="op">)</span>

<span class="co"># Step 2: define the model (design) for the test</span>
<span class="va">model</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/flm_def.html">flm_def</a></span><span class="op">(</span><span class="va">samples</span>, covariate <span class="op">=</span> <span class="st">"sample_type"</span>,
                 numer <span class="op">=</span> <span class="st">"tumor"</span>, denom <span class="op">=</span> <span class="st">"normal"</span>,
                 batch <span class="op">=</span> <span class="st">"sex"</span><span class="op">)</span>

<span class="co"># Step 3: configure the options to run the test, which include the assay that</span>
<span class="co">#         holds the data used to test and the statistical method/framework we</span>
<span class="co">#         should use to perform the test</span>
<span class="va">vdge</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/fdge.html">fdge</a></span><span class="op">(</span><span class="va">model</span>, assay_name <span class="op">=</span> <span class="st">"rnaseq"</span>, method <span class="op">=</span> <span class="st">"voom"</span><span class="op">)</span></code></pre></div>
<p>Perhaps you prefer the <a href="https://f1000research.com/articles/5-1438" class="external-link">edgeR/QLF</a> analysis framework, instead? No problem, we only need to tweak one of the parameters in the last step of the pipeline:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">qdge</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/fdge.html">fdge</a></span><span class="op">(</span><span class="va">model</span>, assay_name <span class="op">=</span> <span class="st">"rnaseq"</span>, method <span class="op">=</span> <span class="st">"edgeR-qlf"</span><span class="op">)</span></code></pre></div>
<p>… or DESeq2, perhaps? No problem, <a href="https://github.com/facilebio/FacileAnalysis/pulls" class="external-link">we accept pull requests</a>!</p>
</div>
<div class="section level2">
<h2 id="interrogation-of-analyses-at-different-levels-of-interactivity">Interrogation of Analyses at Different Levels of Interactivity<a class="anchor" aria-label="anchor" href="#interrogation-of-analyses-at-different-levels-of-interactivity"></a>
</h2>
<p>There are a number of S3 methods a <code>FacileAnalysisResult</code> needs to define in order to be complete. The <code><a href="reference/FacileAnalysisResultViz.html">shine()</a></code>, <code><a href="reference/FacileAnalysisResultViz.html">viz()</a></code>, and <code><a href="reference/FacileAnalysisResultViz.html">report()</a></code> methods allow the end-user to interrogate (and report) the results of an analysis at different levels of interactivity.</p>
<p>Let’s take a look at how these work over a <code>FacileDgeAnalysisResult</code></p>
<div class="section level3">
<h3 id="deep-interaction-via-shine">Deep Interaction via shine()<a class="anchor" aria-label="anchor" href="#deep-interaction-via-shine"></a>
</h3>
<p>The <code>shine(aresult, ...)</code> method provides the richest interactive view over a <code>FacileAnalysisResult</code> by launching a <a href="https://shiny.rstudio.com/articles/gadgets.html" class="external-link">shiny gadget</a> that enables the end-user to fully interrogate the results.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="reference/FacileAnalysisResultViz.html">shine</a></span><span class="op">(</span><span class="va">vdge</span><span class="op">)</span></code></pre></div>
<p><img src="reference/figures/fdge-shine.gif" width="75%"></p>
</div>
<div class="section level3">
<h3 id="interactive-graphics-via-viz">Interactive Graphics via viz()<a class="anchor" aria-label="anchor" href="#interactive-graphics-via-viz"></a>
</h3>
<p>The <code>viz(aresult, ...)</code> methods leverage <a href="https://www.htmlwidgets.org" class="external-link">htmlwidgets</a> to create a JavaScript-powered interactive view of the analysis result, which is detached from a running R-process.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="reference/FacileAnalysisResultViz.html">viz</a></span><span class="op">(</span><span class="va">vdge</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level3">
<h3 id="reporting-results-via-report">Reporting Results via report()<a class="anchor" aria-label="anchor" href="#reporting-results-via-report"></a>
</h3>
<p>While the output of the <code><a href="reference/FacileAnalysisResultViz.html">viz()</a></code> functions can be used directly in Rmarkdown reports, <code>report(aresult, ...)</code> is meant to create a “more complete” (perhaps multi-htmlwidget) view over the result that can be more suitable for inclusion into an Rmarkdown report.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">cap</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span>
  <span class="st">"Differential expression of tumor vs normal samples, in the TCGA bladder"</span>,
  <span class="st">"cancer indication (BLCA)"</span><span class="op">)</span>
<span class="fu"><a href="reference/FacileAnalysisResultViz.html">report</a></span><span class="op">(</span><span class="va">vdge</span>, caption <span class="op">=</span> <span class="va">cap</span><span class="op">)</span></code></pre></div>
<!--
<img src="man/figures/fdge-report.gif" width = "75%" />
-->
</div>
</div>
<div class="section level2">
<h2 id="hybrid-analyses">Hybrid Analyses<a class="anchor" aria-label="anchor" href="#hybrid-analyses"></a>
</h2>
<p>The same differential expression analysis that created the <code>vdge</code> object above can be performed entirely interactively, with the same results.</p>
<p>We can either start the analysis from the same predefined set of samples, but define the linear model and testing framework to use interactively, by launching a the <strong>f</strong>acile <strong>d</strong>ifferential <strong>g</strong>ene <strong>e</strong>xpression gadget:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">vdge2</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/fdge-shiny.html">fdgeGadget</a></span><span class="op">(</span><span class="va">samples</span><span class="op">)</span></code></pre></div>
<p>Or we can perform the whole thing via a GUI which lets us select the subset of samples and run the differential expression analysis without using any code at all:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">vdge3</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/fdge-shiny.html">fdgeGadget</a></span><span class="op">(</span><span class="va">efds</span><span class="op">)</span></code></pre></div>
<p><img src="reference/figures/vignette-fdgeGadget-trim.gif" width="75%"></p>
<p>Assuming the same filtering and testing strategies were selected in the GUI using the <code>fdgeGadget</code> calls above, the objects they return will all be equivalent to the <code>vdge</code> result, which was entirely generated programmatically.</p>
<p><strong>Note</strong>: the screen capture of the <code><a href="reference/fdge-shiny.html">fdgeGadget()</a></code> above actually corresponds to the analyses in the <a href="https://facilebio.github.io/FacileAnalysis/articles/FacileAnalysis-RNAseq.html" class="external-link">RNA-seq the facile way</a> vignette.</p>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>Because the packages in the facileverse have not officially been released yet and they also depend heavily on <a href="https://bioconductor.org" class="external-link">bioconductor</a> packages, we recommend using the <a href="https://cran.r-project.org/package=BiocManager" class="external-link">BiocManager</a> package for installation.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># install.packages("BiocManager")</span>
<span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"facilebio/FacileAnalysis"</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="resources">Resources<a class="anchor" aria-label="anchor" href="#resources"></a>
</h2>
<ul>
<li>
<a href="https://facilebio.github.io/FacileAnalysis" class="external-link">facilebio.github.io/FacileAnalysis</a> (online documentation and vignettes)</li>
<li>
<a href="https://github.com/facilebio/FacileAnalysis/issues" class="external-link">Open an issue</a> (GitHub issues for questions, bug reports, and feature requests)</li>
</ul>
</div>
<div class="section level2">
<h2 id="acknowledgements">Acknowledgements<a class="anchor" aria-label="anchor" href="#acknowledgements"></a>
</h2>
<ul>
<li>Thanks to Morten Just, who developed the <a href="https://github.com/mortenjust/droptogif" class="external-link">droptogif</a>, which we used to convert screen-captured <code>*.mov</code> files to animated gifs.</li>
</ul>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/facilebio/FacileAnalysis/" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/facilebio/FacileAnalysis/issues" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small>Apache License (&gt;= 2.0)</small></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing FacileAnalysis</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Steve Lianoglou <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0002-0924-1754" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>Denali Therapeutics <br><small class="roles"> Copyright holder, funder </small>  </li>
</ul>
</div>

<div class="dev-status">
<h2 data-toc-skip>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/facilebio/FacileAnalysis/actions" class="external-link"><img src="https://github.com/facilebio/FacileAnalysis/workflows/R-CMD-check/badge.svg" alt="R build status"></a></li>
<li><a href="https://www.repostatus.org/#active" class="external-link"><img src="https://www.repostatus.org/badges/latest/active.svg" alt="Project Status: Active"></a></li>
<li><a href="https://www.tidyverse.org/lifecycle/#maturing" class="external-link"><img src="https://img.shields.io/badge/lifecycle-maturing-blue.svg" alt="Lifecycle: Maturing"></a></li>
<li><a href="https://codecov.io/gh/facilebio/FacileAnalysis?branch=master" class="external-link"><img src="https://codecov.io/gh/facilebio/FacileAnalysis/branch/master/graph/badge.svg" alt="Codecov test coverage"></a></li>
</ul>
</div>

  </div>
</div>


      <footer><div class="copyright">
  <p></p>
<p>Developed by Steve Lianoglou, Denali Therapeutics.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.2.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
