<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Construct a bioconductor classed object from an analysis. — biocbox.FacileLinearModelDefinition • FacileAnalysis</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Construct a bioconductor classed object from an analysis. — biocbox.FacileLinearModelDefinition"><meta property="og:description" content="Construct a bioconductor classed object from an analysis."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">FacileAnalysis</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.98.16</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/FacileAnalysis-RNAseq.html">RNA-seq Analysis: The Facile Way</a>
    </li>
    <li>
      <a href="../articles/FacileAnalysis-examples.html">Reproducing genomic anlayses the facile way</a>
    </li>
    <li>
      <a href="../articles/FacileAnalysis-fdge.html">Facile Differential Expression Analysis</a>
    </li>
    <li>
      <a href="../articles/FacileAnalysis-fpca.html">Facile Principal Components Analysis</a>
    </li>
    <li>
      <a href="../articles/FacileAnalysis-fsea.html">Facile (Gene) Set Expression Analysis</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/facilebio/FacileAnalysis/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Construct a bioconductor classed object from an analysis.</h1>
    <small class="dont-index">Source: <a href="https://github.com/facilebio/FacileAnalysis/blob/HEAD/R/biocbox.R" class="external-link"><code>R/biocbox.R</code></a>, <a href="https://github.com/facilebio/FacileAnalysis/blob/HEAD/R/fdge.R" class="external-link"><code>R/fdge.R</code></a></small>
    <div class="hidden name"><code>biocbox.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Construct a bioconductor classed object from an analysis.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="co"># S3 method for FacileLinearModelDefinition</span>
<span class="fu"><a href="https://rdrr.io/pkg/FacileData/man/biocbox.html" class="external-link">biocbox</a></span><span class="op">(</span>
  <span class="va">x</span>,
  assay_name <span class="op">=</span> <span class="cn">NULL</span>,
  method <span class="op">=</span> <span class="cn">NULL</span>,
  features <span class="op">=</span> <span class="cn">NULL</span>,
  filter <span class="op">=</span> <span class="st">"default"</span>,
  filter_universe <span class="op">=</span> <span class="cn">NULL</span>,
  filter_require <span class="op">=</span> <span class="cn">NULL</span>,
  with_sample_weights <span class="op">=</span> <span class="cn">FALSE</span>,
  weights <span class="op">=</span> <span class="cn">NULL</span>,
  block <span class="op">=</span> <span class="cn">NULL</span>,
  prior_count <span class="op">=</span> <span class="fl">0.1</span>,
  <span class="va">...</span>
<span class="op">)</span>

<span class="co"># S3 method for FacileDgeAnalysisResult</span>
<span class="fu"><a href="https://rdrr.io/pkg/FacileData/man/biocbox.html" class="external-link">biocbox</a></span><span class="op">(</span><span class="va">x</span>, cached <span class="op">=</span> <span class="cn">TRUE</span>, <span class="va">...</span><span class="op">)</span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>assay_name</dt>
<dd><p>the name of the assay to pull data for</p></dd>
<dt>method</dt>
<dd><p>the name of the dge method that will be used. This will dictate
the post-processing of the data</p></dd>
<dt>filter</dt>
<dd><p>A filtering policy to remove unintereesting genes.
If <code>"default"</code> (which is the default), then <code><a href="https://rdrr.io/pkg/edgeR/man/filterByExpr.html" class="external-link">edgeR::filterByExpr()</a></code> is
used if we are materializing a <code>DGEList</code>, otherwise lowly expressed
features are removed in a similarly "naive" manner. This can,
alternatively, be a character vector that holds the names of the features
that should be kept. Default value: <code>"default"</code>.</p></dd>
<dt>with_sample_weights</dt>
<dd><p>Some methods that leverage the limma pipeline,
like <code>"voom"</code>, <code>"limma"</code>, and <code>"limma-trend"</code> can leverage sample (array)
quality weights to downweight outlier samples. In the case of
<code>method == "voom"</code>, we use <code><a href="https://rdrr.io/pkg/limma/man/voomWithQualityWeights.html" class="external-link">limma::voomWithQualityWeights()</a></code>, while the
rest use <code><a href="https://rdrr.io/pkg/limma/man/arrayWeights.html" class="external-link">limma::arrayWeights()</a></code>. The choice of <code>method</code> determines which
sample weighting function to sue. Defaults to <code>FALSE</code>.</p></dd>
<dt>prior_count</dt>
<dd><p>The pseudo-count to add to count data. Used primarily
when running the <code>limma-trend</code> method on count (RNA-seq) data.</p></dd>
<dt>...</dt>
<dd><p>passed down to internal modeling and filtering functions.</p></dd>
<dt>sample_info</dt>
<dd><p>a <code>facile_frame</code> that enumerates the samples to fetch
data for, as well as the covariates used in downstream analysis</p></dd>
</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>a DGEList or EList with assay data in the correct place, and all of
the covariates in the <code>$samples</code> or <code>$targerts</code> data.frame that are requied
to test the model in <code>mdef</code>.</p>
    </div>
    <div id="linear-model-definitions">
    <h2>Linear Model Definitions</h2>
    

<p>This function accepts a model defined using using <code><a href="flm_def.html">flm_def()</a></code> and
creates the appropriate Bioconductor assay container to test the model
given the <code>assay_name</code> and dge <code>method</code> specified by the user.</p>
<p>This function currently supports retrieving data and whipping it into
a DGEList (for count-like data) and an EList for data that can be analyzed
with one form limma or another.</p>
<p>Assumptions on different <code>assay_type</code> values include:</p><ul><li><p><code>rnaseq</code>: assumed to be "vanilla" bulk rnaseq gene counts</p></li>
<li><p><code>umi</code>: data from bulk rnaseq, UMI data, like quantseq</p></li>
<li><p><code>tpm</code>: TPM values. These will be <code>log2(TPM + prior_count)</code> transformed,
then differentially tested using the limma-trended pipeline</p></li>
</ul><p>TODO: support affymrna, affymirna, etc. assay types</p>
<p>The "filter" parameters are described in the <code><a href="fdge.html">fdge()</a></code> function for now.</p>
    </div>
    <div id="faciledgeanalysisresult">
    <h2>FacileDgeAnalysisResult</h2>
    

<p>Given a FacileDgeAnalysisResult, we can re-materialize the Bioconductor assay
container used within the differential testing pipeline used from <code><a href="fdge.html">fdge()</a></code>.
Currently we have limited our analysis framework to either work over DGEList
(edgeR) or EList (limma) containers.</p>
    </div>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Steve Lianoglou, Denali Therapeutics.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.2.</p>
</div>

      </footer></div>

  


  

  </body></html>

