<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Construct a bioconductor classed object from an analysis. — biocbox.FacileLinearModelDefinition • FacileAnalysis</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



  <link href="../extra.css" rel="stylesheet">
  

<meta property="og:title" content="Construct a bioconductor classed object from an analysis. — biocbox.FacileLinearModelDefinition" />
<meta property="og:description" content="Construct a bioconductor classed object from an analysis." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">FacileAnalysis</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.98.5</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/FacileAnalysis-RNAseq.html">RNA-seq Analysis: The Facile Way</a>
    </li>
    <li>
      <a href="../articles/FacileAnalysis-examples.html">Reproducing genomic anlayses the facile way</a>
    </li>
    <li>
      <a href="../articles/FacileAnalysis-fdge.html">Facile Differential Expression Analysis</a>
    </li>
    <li>
      <a href="../articles/FacileAnalysis-fpca.html">Facile Principal Components Analysis</a>
    </li>
    <li>
      <a href="../articles/FacileAnalysis-fsea.html">Facile (Gene) Set Expression Analysis</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/facilebio/FacileAnalysis/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Construct a bioconductor classed object from an analysis.</h1>
    <small class="dont-index">Source: <a href='https://github.com/facilebio/FacileAnalysis/blob/master/R/biocbox.R'><code>R/biocbox.R</code></a>, <a href='https://github.com/facilebio/FacileAnalysis/blob/master/R/fdge.R'><code>R/fdge.R</code></a></small>
    <div class="hidden name"><code>biocbox.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Construct a bioconductor classed object from an analysis.</p>
    </div>

    <pre class="usage"><span class='co'># S3 method for FacileLinearModelDefinition</span>
<span class='fu'><a href='https://rdrr.io/pkg/FacileData/man/biocbox.html'>biocbox</a></span><span class='op'>(</span>
  <span class='va'>x</span>,
  assay_name <span class='op'>=</span> <span class='cn'>NULL</span>,
  method <span class='op'>=</span> <span class='cn'>NULL</span>,
  features <span class='op'>=</span> <span class='cn'>NULL</span>,
  filter <span class='op'>=</span> <span class='st'>"default"</span>,
  filter_universe <span class='op'>=</span> <span class='cn'>NULL</span>,
  filter_require <span class='op'>=</span> <span class='cn'>NULL</span>,
  with_sample_weights <span class='op'>=</span> <span class='cn'>FALSE</span>,
  weights <span class='op'>=</span> <span class='cn'>NULL</span>,
  block <span class='op'>=</span> <span class='cn'>NULL</span>,
  prior_count <span class='op'>=</span> <span class='fl'>0.1</span>,
  <span class='va'>...</span>
<span class='op'>)</span>

<span class='co'># S3 method for FacileDgeAnalysisResult</span>
<span class='fu'><a href='https://rdrr.io/pkg/FacileData/man/biocbox.html'>biocbox</a></span><span class='op'>(</span><span class='va'>x</span>, <span class='va'>...</span><span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>assay_name</th>
      <td><p>the name of the assay to pull data for</p></td>
    </tr>
    <tr>
      <th>method</th>
      <td><p>the name of the dge method that will be used. This will dictate
the post-processing of the data</p></td>
    </tr>
    <tr>
      <th>filter</th>
      <td><p>A filtering policy to remove unintereesting genes.
If <code>"default"</code> (which is the default), then <code><a href='https://rdrr.io/pkg/edgeR/man/filterByExpr.html'>edgeR::filterByExpr()</a></code> is
used if we are materializing a <code>DGEList</code>, otherwise lowly expressed
features are removed in a similarly "naive" manner. This can,
alternatively, be a character vector that holds the names of the features
that should be kept. Default value: <code>"default"</code>.</p></td>
    </tr>
    <tr>
      <th>with_sample_weights</th>
      <td><p>Some methods that leverage the limma pipeline,
like <code>"voom"</code>, <code>"limma"</code>, and <code>"limma-trend"</code> can leverage sample (array)
quality weights to downweight outlier samples. In the case of
<code>method == "voom"</code>, we use <code><a href='https://rdrr.io/pkg/limma/man/voomWithQualityWeights.html'>limma::voomWithQualityWeights()</a></code>, while the
rest use <code><a href='https://rdrr.io/pkg/limma/man/arrayWeights.html'>limma::arrayWeights()</a></code>. The choice of <code>method</code> determines which
sample weighting function to sue. Defaults to <code>FALSE</code>.</p></td>
    </tr>
    <tr>
      <th>prior_count</th>
      <td><p>The pseudo-count to add to count data. Used primarily
when running the <code>limma-trend</code> method on count (RNA-seq) data.</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>passed down to internal modeling and filtering functions.</p></td>
    </tr>
    <tr>
      <th>sample_info</th>
      <td><p>a <code>facile_frame</code> that enumerates the samples to fetch
data for, as well as the covariates used in downstream analysis</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>a DGEList or EList with assay data in the correct place, and all of
the covariates in the <code>$samples</code> or <code>$targerts</code> data.frame that are requied
to test the model in <code>mdef</code>.</p>
    <h2 class="hasAnchor" id="linear-model-definitions"><a class="anchor" href="#linear-model-definitions"></a>Linear Model Definitions</h2>

    

<p>This function accepts a model defined using using <code><a href='flm_def.html'>flm_def()</a></code> and
creates the appropriate Bioconductor assay container to test the model
given the <code>assay_name</code> and dge <code>method</code> specified by the user.</p>
<p>This function currently supports retrieving data and whipping it into
a DGEList (for count-like data) and an EList for data that can be analyzed
with one form limma or another.</p>
<p>Assumptions on different <code>assay_type</code> values include:</p><ul>
<li><p><code>rnaseq</code>: assumed to be "vanilla" bulk rnaseq gene counts</p></li>
<li><p><code>umi</code>: data from bulk rnaseq, UMI data, like quantseq</p></li>
<li><p><code>tpm</code>: TPM values. These will be <code><a href='https://rdrr.io/r/base/Log.html'>log2(TPM + prior_count)</a></code> transformed,
then differentially tested using the limma-trended pipeline</p></li>
</ul>

<p>TODO: support affymrna, affymirna, etc. assay types</p>
<p>The "filter" parameters are described in the <code><a href='fdge.html'>fdge()</a></code> function for now.</p>
    <h2 class="hasAnchor" id="faciledgeanalysisresult"><a class="anchor" href="#faciledgeanalysisresult"></a>FacileDgeAnalysisResult</h2>

    

<p>Given a FacileDgeAnalysisResult, we can re-materialize the Bioconductor assay
container used within the differential testing pipeline used from <code><a href='fdge.html'>fdge()</a></code>.
Currently we have limited our analysis framework to either work over DGEList
(edgeR) or EList (limma) containers.</p>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Steve Lianoglou, Denali Therapeutics.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


