<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Performs Feature (Gene) Set Enrichment Analyses — ffsea • FacileAnalysis</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Performs Feature (Gene) Set Enrichment Analyses — ffsea"><meta property="og:description" content='Currently we support running a number of feature (gene) set enrichment
analyses downstream of a some other FacileAnalysisResult (ie.,
[fdge() | fpca()] %&amp;gt;% ffsea()), or over an arbitrary data.frame of
feature-level statistics.
Please refer to the examples here, as well as theh "Feature Set Analysis"
section of the vignette for more information.'><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">FacileAnalysis</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.98.16</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/FacileAnalysis-RNAseq.html">RNA-seq Analysis: The Facile Way</a>
    </li>
    <li>
      <a href="../articles/FacileAnalysis-examples.html">Reproducing genomic anlayses the facile way</a>
    </li>
    <li>
      <a href="../articles/FacileAnalysis-fdge.html">Facile Differential Expression Analysis</a>
    </li>
    <li>
      <a href="../articles/FacileAnalysis-fpca.html">Facile Principal Components Analysis</a>
    </li>
    <li>
      <a href="../articles/FacileAnalysis-fsea.html">Facile (Gene) Set Expression Analysis</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/facilebio/FacileAnalysis/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Performs Feature (Gene) Set Enrichment Analyses</h1>
    <small class="dont-index">Source: <a href="https://github.com/facilebio/FacileAnalysis/blob/HEAD/R/ffsea.R" class="external-link"><code>R/ffsea.R</code></a></small>
    <div class="hidden name"><code>ffsea.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Currently we support running a number of feature (gene) set enrichment
analyses downstream of a <em>some other</em> <code>FacileAnalysisResult</code> (ie.,
<code>[fdge() | fpca()] %&gt;% ffsea()</code>), or over an arbitrary data.frame of
feature-level statistics.</p>
<p>Please refer to the examples here, as well as theh <em>"Feature Set Analysis"</em>
section of the vignette for more information.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="fu">ffsea</span><span class="op">(</span><span class="va">x</span>, <span class="va">fsets</span>, methods <span class="op">=</span> <span class="cn">NULL</span>, <span class="va">...</span><span class="op">)</span>

<span class="co"># S3 method for FacileFseaAnalysisResult</span>
<span class="fu">result</span><span class="op">(</span><span class="va">x</span>, name <span class="op">=</span> <span class="st">"object"</span>, <span class="va">...</span><span class="op">)</span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>x</dt>
<dd><p>A <code>FacileAnalysisResult</code> object, or a data.frame with feature-level
statistics, minimally with a <code>"feature_id"</code> column as well as one or more
<code>numeric</code> columns to rank features on.</p></dd>
<dt>fsets</dt>
<dd><p>The feature sets (likely genesets) to use for testing. This
object will be passed through the <code><a href="https://rdrr.io/pkg/sparrow/man/GeneSetDb-class.html" class="external-link">sparrow::GeneSetDb()</a></code> constructor to
create a <code>GeneSetDb</code> object that will be used for testing.</p></dd>
<dt>methods</dt>
<dd><p>A character vector of GSEA methods to use on <code>x</code>. Chose any
of the <code>method</code> names listed by running <code>ffsea_methods(x)</code>. If <code>NULL</code>
(default), the first method from  <code>ffsea_methods(x)</code> will be selected.</p></dd>
</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>A FacileFseaAnalysisResult object, which includes a <code>SparrowResult</code>object as it's <code>result()</code>. The geneset level statistics for each of the
methods that were run are available via <code>tidy(ffsea.res, "&lt;method_name&gt;")</code>.</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p>:
When running <code>ffsea</code> over a <code>FacileAnalysisResult</code>, the types of methods
that can be run, and their configuration are preconfigured with reaonable
defaults.</p>
<p>When providing a generic <code>data.frame</code> of feature-level statistics to run
enrichment tests over, the user has to specificy a few more parmaters.
If running a "pre-ranked" test, (<code>method %in% c("cameraPR", "fgsea")</code>) the
name of a numeric column in <code>x</code> must be specified to rank the features by,
and the order by which to do that using the <code>rank_by</code> and <code>rank_order</code>
arguments, respectively.</p>
<p>If running an overrepresentation analysis-style test (<code>method = "ora"</code>),
the user must specifcy the name of a logical column that indicates
(when <code>TRUE</code>) that a feature should be included for enrichment testing. The
user can optionally specify a <code>group_by</code> column, like <code>"direction"</code>, that
will be used to split the selected features into groups to perform more
specific enrichmen tests. This allows you enrichment tests to be run
separately for <code>"up"</code> and <code>"down"</code> regulated genes separately, for example.</p>
<p>Lastly, the user can provide the name of another numeric column in <code>x</code> with
<code>biased_by</code> which can be used to account for bias in the enrichment tests,
such as gene length, GC content, etc.</p>
<p>Gene sets must be supplied as a <code><a href="https://rdrr.io/pkg/sparrow/man/GeneSetDb-class.html" class="external-link">sparrow::GeneSetDb()</a></code> object.</p>
    </div>
    <div id="gsea-methods">
    <h2>GSEA Methods</h2>
    

<p>Currently, only the following GSEA methods are supported:</p><ul><li><p><code>"cameraPR"</code>: Delegates to <code><a href="https://rdrr.io/pkg/limma/man/camera.html" class="external-link">limma::cameraPR()</a></code> to perform a competitive
gene set test based on feature ranks imposed downstream of an analysis</p></li>
<li><p><code>"fgsea"</code>: Delegates to <code><a href="https://rdrr.io/pkg/fgsea/man/fgsea.html" class="external-link">fgsea::fgsea()</a></code> to perform another version of
a competitive gene set test based on ranks.</p></li>
<li><p><code>"ora"</code>: Performs an overrepresentation analysis test. The user
must specify the name of <code>logical</code> column (<code>select_by</code>) from the input
which is used to indicate the features that are selected for enrichment
analysis. The user can optionally provide the name of a <code>numeric</code> column
(<code>biased_by</code>) and <code>character</code> column (<code>group_by</code>), which will adjust the
enrichment test for a covariate that may induce a bias in the DGE results,
and also run follow up enrichment tests based by differnt groups of
features (<code>group_by</code>). For example, the result table might have a
<code>"direction"</code> column, which specifies the direciton of differential
expression (<code>"up"</code>, or <code>"down"</code>). In this case, enrichment tests will be
run over <em>all</em> features together, and then independantly for the ones that
are <code>"up"</code>, and <code>"down"</code>.</p></li>
</ul></div>
    <div id="gsea-statistics">
    <h2>GSEA Statistics</h2>
    

<p>The geneset level statistics can be extracted from the
<code>FacileFseaAnalysisResult</code> on a per-method basis usig the <code><a href="https://rdrr.io/pkg/generics/man/tidy.html" class="external-link">tidy()</a></code> function.
For instance, if <code>ffsea()</code> was called with
<code>fres &lt;- ffsea(..., methods = c("cameraPR", "ora")</code>, the <code>"cameraPR"</code>
results can be extracted via <code>tidy(fres, "cameraPR")</code></p>
    </div>
    <div id="development-notes">
    <h2>Development Notes</h2>
    

<p>This functionality delegates to <code><a href="https://rdrr.io/pkg/sparrow/man/seas.html" class="external-link">sparrow::seas()</a></code> to do all of the work. The
sparrow::seas interface is undergoing a bit of refactoring in order to better
support a table of feature statistcs as input (for preranked and enrichment
tests), so the <code>"methods"</code> supported via <code>ffsea()</code> are limited to a subset
of the ones wrapped by <code><a href="https://rdrr.io/pkg/sparrow/man/seas.html" class="external-link">sparrow::seas()</a></code>, as enumerated below.</p>
    </div>
    <div id="accessing-results">
    <h2>Accessing Results</h2>
    

<p>We are in a bit of a schizophrenic state right now, where <code><a href="https://rdrr.io/pkg/generics/man/tidy.html" class="external-link">tidy()</a></code> is
being the de-facto way to answer "tidy" like results (instead of result()).</p>
<p>This is not to say that <code>result()</code> can't also return something that's
"tidy", but in this case, result(ffsea.result) will return the
SparrowResult object itself, and <code>tidy(ffsea.result)</code> will dispatch
to <code><a href="https://rdrr.io/pkg/sparrow/man/results.html" class="external-link">sparrow::result()</a></code> to fetch the gsea statistcs for the method
requested.</p><div class="sourceCode"><pre><code><span class="va">mgres</span> <span class="op">&lt;-</span> <span class="fu">result</span><span class="op">(</span><span class="va">ffsea.res</span><span class="op">)</span> <span class="co"># returns the SparrowResult object</span>
<span class="va">camera.stats</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/generics/man/tidy.html" class="external-link">tidy</a></span><span class="op">(</span><span class="va">ffsea.res</span>, name <span class="op">=</span> <span class="st">"cameraPR"</span><span class="op">)</span></code></pre></div>

    </div>
    <div id="see-also">
    <h2>See also</h2>
    <div class="dont-index"><p>https://github.com/lianos/sparrow</p></div>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span class="va">gdb</span> <span class="op">&lt;-</span> <span class="fu">sparrow</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/sparrow/man/examples.html" class="external-link">exampleGeneSetDb</a></span><span class="op">(</span><span class="op">)</span></span>
<span class="r-in"><span class="va">efds</span> <span class="op">&lt;-</span> <span class="fu">FacileData</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/FacileData/man/exampleFacileDataSet.html" class="external-link">exampleFacileDataSet</a></span><span class="op">(</span><span class="op">)</span></span>
<span class="r-in"></span>
<span class="r-in"><span class="co"># GSEA from t-test result ---------------------------------------------------</span></span>
<span class="r-in"><span class="va">ttest.res</span> <span class="op">&lt;-</span> <span class="va">efds</span> <span class="op"><a href="https://rdrr.io/pkg/magrittr/man/pipe.html" class="external-link">%&gt;%</a></span></span>
<span class="r-in">  <span class="fu">FacileData</span><span class="fu">::</span><span class="fu">filter_samples</span><span class="op">(</span><span class="va">indication</span> <span class="op">==</span> <span class="st">"CRC"</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/pkg/magrittr/man/pipe.html" class="external-link">%&gt;%</a></span></span>
<span class="r-in">  <span class="fu"><a href="flm_def.html">flm_def</a></span><span class="op">(</span>covariate <span class="op">=</span> <span class="st">"sample_type"</span>, numer <span class="op">=</span> <span class="st">"tumor"</span>, denom <span class="op">=</span> <span class="st">"normal"</span>,</span>
<span class="r-in">          batch <span class="op">=</span> <span class="st">"sex"</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/pkg/magrittr/man/pipe.html" class="external-link">%&gt;%</a></span></span>
<span class="r-in">  <span class="fu"><a href="fdge.html">fdge</a></span><span class="op">(</span>method <span class="op">=</span> <span class="st">"voom"</span><span class="op">)</span></span>
<span class="r-in"></span>
<span class="r-in"><span class="va">ttest.gsea</span> <span class="op">&lt;-</span> <span class="fu">ffsea</span><span class="op">(</span><span class="va">ttest.res</span>, <span class="va">gdb</span>, methods <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"cameraPR"</span>, <span class="st">"ora"</span><span class="op">)</span>,</span>
<span class="r-in">                    biased_by <span class="op">=</span> <span class="st">"effective_length"</span><span class="op">)</span></span>
<span class="r-in"><span class="kw">if</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/interactive.html" class="external-link">interactive</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span class="r-in">  <span class="fu"><a href="FacileAnalysisResultViz.html">viz</a></span><span class="op">(</span><span class="va">ttest.gsea</span>, type <span class="op">=</span> <span class="st">"density"</span>, name <span class="op">=</span> <span class="st">"HALLMARK_HEDGEHOG_SIGNALING"</span><span class="op">)</span></span>
<span class="r-in">  <span class="fu"><a href="FacileAnalysisResultViz.html">viz</a></span><span class="op">(</span><span class="va">ttest.gsea</span>, type <span class="op">=</span> <span class="st">"gsea"</span>, name <span class="op">=</span> <span class="st">"HALLMARK_HEDGEHOG_SIGNALING"</span><span class="op">)</span></span>
<span class="r-in"></span>
<span class="r-in">  <span class="fu"><a href="FacileAnalysisResultViz.html">shine</a></span><span class="op">(</span><span class="va">ttest.igsea</span><span class="op">)</span></span>
<span class="r-in">  <span class="va">ttest.igsea</span> <span class="op">&lt;-</span> <span class="fu"><a href="interactive-ffsea.html">ffseaGadget</a></span><span class="op">(</span><span class="va">ttest.res</span>, <span class="va">gdb</span><span class="op">)</span></span>
<span class="r-in"><span class="op">}</span></span>
<span class="r-in"></span>
<span class="r-in"><span class="va">camera.stats</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/generics/man/tidy.html" class="external-link">tidy</a></span><span class="op">(</span><span class="va">ttest.gsea</span>, <span class="st">"cameraPR"</span><span class="op">)</span></span>
<span class="r-in"><span class="va">ora.stats</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/generics/man/tidy.html" class="external-link">tidy</a></span><span class="op">(</span><span class="va">ttest.gsea</span>, <span class="st">"ora"</span><span class="op">)</span></span>
<span class="r-in"></span>
<span class="r-in"><span class="co"># GSEA from ANOVA result ----------------------------------------------------</span></span>
<span class="r-in"><span class="va">stage.anova</span> <span class="op">&lt;-</span> <span class="va">efds</span> <span class="op"><a href="https://rdrr.io/pkg/magrittr/man/pipe.html" class="external-link">%&gt;%</a></span></span>
<span class="r-in">  <span class="fu">FacileData</span><span class="fu">::</span><span class="fu">filter_samples</span><span class="op">(</span><span class="va">indication</span> <span class="op">==</span> <span class="st">"BLCA"</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/pkg/magrittr/man/pipe.html" class="external-link">%&gt;%</a></span></span>
<span class="r-in">  <span class="fu"><a href="flm_def.html">flm_def</a></span><span class="op">(</span>covariate <span class="op">=</span> <span class="st">"stage"</span>, batch <span class="op">=</span> <span class="st">"sex"</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/pkg/magrittr/man/pipe.html" class="external-link">%&gt;%</a></span></span>
<span class="r-in">  <span class="fu"><a href="fdge.html">fdge</a></span><span class="op">(</span>method <span class="op">=</span> <span class="st">"voom"</span><span class="op">)</span></span>
<span class="r-in"><span class="va">anova.gsea</span> <span class="op">&lt;-</span> <span class="fu">ffsea</span><span class="op">(</span><span class="va">stage.anova</span>, <span class="va">gdb</span><span class="op">)</span></span>
<span class="r-in"><span class="kw">if</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/interactive.html" class="external-link">interactive</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span class="r-in"> <span class="co"># TODO: shine(anova.gsea) doesn't work</span></span>
<span class="r-in"> <span class="fu"><a href="FacileAnalysisResultViz.html">shine</a></span><span class="op">(</span><span class="va">anova.gsea</span><span class="op">)</span></span>
<span class="r-in"> <span class="co"># We can generate the same GSEA result like so</span></span>
<span class="r-in"> <span class="va">anova.gsea2</span> <span class="op">&lt;-</span> <span class="fu"><a href="interactive-ffsea.html">ffseaGadget</a></span><span class="op">(</span><span class="va">stage.anova</span>, gdb <span class="op">=</span> <span class="va">gdb</span><span class="op">)</span></span>
<span class="r-in"><span class="op">}</span></span>
<span class="r-in"><span class="co"># GSEA over loadings on a Principal Component -------------------------------</span></span>
<span class="r-in"><span class="va">pca.crc</span> <span class="op">&lt;-</span> <span class="va">efds</span> <span class="op"><a href="https://rdrr.io/pkg/magrittr/man/pipe.html" class="external-link">%&gt;%</a></span></span>
<span class="r-in">  <span class="fu">FacileData</span><span class="fu">::</span><span class="fu">filter_samples</span><span class="op">(</span><span class="va">indication</span> <span class="op">==</span> <span class="st">"CRC"</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/pkg/magrittr/man/pipe.html" class="external-link">%&gt;%</a></span></span>
<span class="r-in">  <span class="fu"><a href="fpca.html">fpca</a></span><span class="op">(</span><span class="op">)</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required namespace: testthat</span>
<span class="r-in"><span class="va">pca1.gsea</span> <span class="op">&lt;-</span> <span class="fu">ffsea</span><span class="op">(</span><span class="va">pca.crc</span>, <span class="va">gdb</span>, dim <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Fraction of gene set IDs that match rownames in the expression object are low: 4.22% </span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Deactivating 18 gene sets because conformation of GeneSetDb to the target creates gene sets smaller than 2 or greater than Inf</span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Steve Lianoglou, Denali Therapeutics.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.2.</p>
</div>

      </footer></div>

  


  

  </body></html>

